From 4c8ef0cd06e6ddea1afc2dbb6ef2a4b6d3c84259 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Sun, 22 Jun 2025 18:08:34 +0200
Subject: [PATCH] HACK: winex11: Skip delay on KDE if wm_state is unchanged

Attempt at fixing an edge-case where a black window is created upon alt-tab on KDE when the window state is unchanged.
Can be disabled with WINE_DISABLE_KDE_HACKS=1.
---
 dlls/winex11.drv/window.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 726478ff040..2368424111a 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -935,10 +935,24 @@ static BOOL window_needs_net_wm_state_change_delay( struct x11drv_win_data *data
     return (data->desired_state.net_wm_state ^ data->pending_state.net_wm_state) & config_notify_mask;
 }
 
+static BOOL enable_kde_hacks(struct x11drv_win_data *data) {
+    static char cached = -1;
+    if (cached == -1) {
+        cached = 0;
+        const char *e = getenv("WINE_DISABLE_KDE_HACKS");
+        if (wm_is_kde(data->display) && (!e || strcmp(e, "1") != 0)) {
+            FIXME("KDE detected, enabling focus fixes..\n");
+            cached = 1;
+        }
+    }
+    return !!cached;
+}
+
 static BOOL window_needs_config_change_delay( struct x11drv_win_data *data )
 {
     if (!data->managed || data->embedded) return FALSE; /* window is not managed or is embedded, safe to make changes */
     if (data->pending_state.wm_state == WithdrawnState) return FALSE; /* window is unmapped, should be safe to make any change */
+    if (data->current_state.wm_state == data->pending_state.wm_state && enable_kde_hacks(data)) return FALSE; /* skip delay for KDE if window state isn't changing */
     if (data->configure_serial) return TRUE; /* another config update is pending, wait for it to complete */
     /* delay any config request when a _NET_WM_STATE or _MOTIF_WM_HINTS change which might trigger a ConfigureNotify is in flight */
     return (data->net_wm_state_serial && (data->pending_state.net_wm_state ^ data->current_state.net_wm_state) & config_notify_mask) ||
-- 
2.50.0

